docs/connections.md

```diff
@@ -0,0 +1,597 @@
+and operations. The purpose is not to specify the whole language. Just enough
+for IDE team members to be able to reason where identifiers are introduced and
+what entity identifier usage refers to.
+
+This is the base allowing IDE to describe what connections are in the displayed
+graph. The covered topics are mostly identifiers, scopes and their interactions.
+
+
+## Identifier
+Identifier is a name that denotes a value, i.e. that is bound to a value. The
+compiler also keeps track of its type, i.e. the set of possible values. Syntactically we recognize:
+* variable names, being names that do not contain upper-cased characters, like `foo2`
+  or `make_new`;
+* constructor names, which are like variables but with first character and every
+  character directly following underscore upper-cased (e.g. `Foo2` or
+  `Make_New`);
+* operator names, being built solely from operator symbols (e.g. `+` or `<$>`).
+  Specifically, an operator name may contain only characters from the following
+  set: `!$%&*+-/<>?^~|:\,.()[]{}=`. Not every sequence of these characters is a valid operator name, as they can collide with other language constructs.
+  
+Any other names that do not match requirements above, like `HTTP`, `foO` or
+`Make_new` are not allowed and their behavior will not be specified.
+
+In non-pattern contexts, referring to the existing identifier is
+case-insensitive. So `foo` can be referred to either as `foo` or `Foo` with no
+difference.  In pattern context, lower-cased names are used to introduce a
+binding (or a constraint), while upper-cased name will refer to an already bound
+identifier. In short, variable name is allowed to shadow, while constructor 
+always unambiguously refers to an externally visible identifier.
+
+Operator names behave either as:
+* variable names when placed in a prefix position (e.g. `+` in `+ ab`);
+* or as constructor names when placed in an infix position (e.g. `,` in `a,b`).
+
+Number and text literals (like `5` or `"Hello"`) are treated as
+constructor names.
+
+Identifiers can be introduced by a binding (e.g. using assignment or lambda
+argument matching) or when adding type constraints (using type ascriptions). 
+
+## Scope
+Scope is the span in the code which shares the identifiers set of visible,
+available identifiers. Scopes are a span-tree structure, covering the whole program code.
+
+Nested scope is allowed to:
+* access identifiers from the outer scope;
+* shadow identifiers from nested scope with a new binding;
+* introduce new constraints on the identifiers from parent (or own) scopes.
+
+The same identifier may be bound multiple times in the same scope (overloading).
+It is allowed only for method overloads that differ in the type of the `this`
+parameter. This limitation may be relaxed in the future, if proper motivating use-cases are found.
+
+The identifier is always accessible scope-wide, both before and after the line
```

Comment:
This isn't something that is (or can be) handled at the moment. Additionally, it doesn't make sense. The definition of the desugaring of monadic contexts (in terms of bind) introduces an ordering dependency. If we want to distinguish between expressions `in Pure` and lift those into mutually recursive let blocks, that is a whole other kettle of fish that will require the typechecker.

Reply:
Yeah we definitely want it to work eventually, but this is very much _eventually_.